graphql_compiler/schema_transformation/utils.py

```diff
@@ -39,16 +40,54 @@ class InvalidTypeNameError(SchemaTransformError):
     """
 
 
-class SchemaNameConflictError(SchemaTransformError):
-    """Raised when renaming or merging types or fields cause name conflicts.
+class SchemaMergeNameConflictError(SchemaTransformError):
+    """Raised when merging types or fields cause name conflicts.
 
-    This may be raised if a field or type is renamed to conflict with another field or type,
-    if two merged schemas share an identically named field or type, or if a
+    This may be raised if two merged schemas share an identically named field or type, or if a
     CrossSchemaEdgeDescriptor provided when merging schemas has an edge name that causes a
     name conflict with an existing field.
     """
 
 
+class SchemaRenameNameConflictError(SchemaTransformError):
+    """Raised when renaming types or fields cause name conflicts.
+
+    This may be raised if a field or type is renamed to conflict with another field or type.
+    """
+
+    name_conflicts: Dict[str, Set[str]]
+    renamed_to_scalar_conflicts: Dict[str, str]
+
+    def __init__(
+        self, name_conflicts: Dict[str, Set[str]], renamed_to_scalar_conflicts: Dict[str, str]
+    ) -> None:
+        """Record all renaming conflicts."""
+        super().__init__()
+        self.name_conflicts = name_conflicts
+        self.renamed_to_scalar_conflicts = renamed_to_scalar_conflicts
+
+    def __str__(self) -> str:
+        """Explain renaming conflict and the fix."""
+        name_conflicts_message = ""
+        if self.name_conflicts:
+            name_conflicts_message = (
+                f"Applying the renaming would produce a schema in which multiple types have the "
+                f"same name, which is an illegal schema state. The name_conflicts dict describes "
+                f"these problems: {self.name_conflicts}. For each key k in name_conflicts, "
+                f"name_conflicts[k] is the set of types in the original schema that get mapped to "
+                f"k in the new schema. To fix this, modify the renamings argument of rename_schema "
+                f"to ensure that no two types in the renamed schema have the same name."
+            )
+        renamed_to_scalar_conflicts_message = ""
+        if self.renamed_to_scalar_conflicts:
+            renamed_to_scalar_conflicts_message = (
+                f"Applying the renaming would rename the following types to a name already used by "
+                f"a scalar: {self.renamed_to_scalar_conflicts}. To fix this, ensure that no type "
+                f"name is mapped to a scalar's name."
+            )
+        return " ".join(filter(None, [name_conflicts_message, renamed_to_scalar_conflicts_message]))
```

Comment:
Is `" ".join([])` AKA empty string a legal output here? If not, let's prevent this error object from being constructed in such a way, and add an assertion against it here.

This is another thing that a test could have potentially uncovered.